### PR TITLE
New dependency manager

### PR DESCRIPTION
**Pull request by joto**
The new dependency manager takes over some of the responsibilities of the middle. This way we decouple the dependency tracking from the object storage. It makes the code easier to understand and to change.

There are more refactorings coming after this.

### CODE REVIEW

File: src/dependency-manager.hpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+public:
+    virtual ~dependency_manager_t() = default;
+
+    /**
+     * Mark a node as changed to trigger the propagation of this change to
+     * ways and relations.
+     *
+     * This has to be called *after* the object was stored in the object store.
+     */
+    virtual void node_changed(osmid_t) {};
```
Comment (lonvia) [inline src/dependency-manager.hpp:43]: remove semicolon, also below

File: src/dependency-manager.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    for (auto const rel_id : m_object_store->get_rels_by_way(id)) {
+        m_rels_pending_tracker.mark(rel_id);
+    }
+}
+
+void full_dependency_manager_t::relation_changed(osmid_t id) {
+    if (m_rels_pending_tracker.is_marked(id)) {
+        return;
```
Comment (lonvia) [inline src/dependency-manager.cpp:27]: This is not true. A marked relation has not necessarily marked its dependencies. That would only be true if we called relation_changed in lines 11, 21 and 31.

File: src/osm2pgsql.cpp [hunk 3]
```diff
@@ -63,7 +63,17 @@ int main(int argc, char *argv[])
...
+            std::any_of(outputs.cbegin(), outputs.cend(),
+                        [](std::shared_ptr<output_t> const &output) {
+                            return output->need_forward_dependencies();
+                        });
+
+        auto dependency_manager = std::unique_ptr<dependency_manager_t>(
+            need_dependencies ? new full_dependency_manager_t{middle.get()}
+                              : new dependency_manager_t{});
+
+        osmdata_t osmdata{dependency_manager.get(), middle, outputs};
```
Comment (lonvia) [inline src/osm2pgsql.cpp:76]: Any particular reason not to hand ownership over to osmdata? Or rather move the code into the osmdata constructor?

File: src/dependency-manager.hpp [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    template <typename TOutputIterator>
+    void get_pending_relation_ids(TOutputIterator &&it) {
+        osmid_t id;
+        while (id_tracker::is_valid(id = m_rels_pending_tracker.pop_mark())) {
+            *it++ = id;
+        }
+    }
+
+private:
+    middle_t* m_object_store;
```
Comment (lonvia) [inline src/dependency-manager.hpp:155]: make that a shared ptr?

Comment (joto) [reply src/dependency-manager.hpp:155]: Bare pointer is all we need here and we know it outlives the dependency manager.

File: src/osm2pgsql.cpp (hunk 3 above)
Comment (joto) [reply src/osm2pgsql.cpp:76]: Currently the dependency manager doesn't need any configuration. But if it does in the future, so I think it is better created here where the options are available.

Moving into osmdata would be possible, just as with middle and outputs. That's certainly something to consider, but I'd rather do all of them or none of them.

File: src/dependency-manager.hpp (hunk 1 above)
Comment (joto) [reply src/dependency-manager.hpp:43]: fixed

File: src/dependency-manager.cpp (hunk 2 above)
Comment (joto) [reply src/dependency-manager.cpp]: fixed

File: src/dependency-manager.hpp (hunk 4 above)
Comment (joto) [reply src/dependency-manager.hpp]: I added a commit that uses a share_ptr here.

File: src/osm2pgsql.cpp (hunk 3 above)
Comment (joto) [reply src/osm2pgsql.cpp]: I added a commit that moves the dependency manager into osmdata.

**Pull request merged**